keg_auth/libs/authenticators.py

```diff
@@ -966,3 +975,136 @@ def get_authenticated_user(self):
+
+
+class PasswordPolicy:
+    """
+    A base class that defines password requirements for the application.
+    This class defines some basic, common validations and can be extended or limited by subclassing.
+
+    To define additional password checks, create a method on your subclass that accepts a
+    password string and a user entity object and raises PasswordPolicyError if the password does
+    not meet the requirement you intend to check. Then override password_checks to add your method
+    to the returned list of methods.
+
+    To remove a password check that is enabled by default, override password_checks and return only
+    the methods you wish to use.
+
+    Default settings are based on NIST guidelines and some common restrictions.
+    """
+
+    """
+    Minimum password length for check_length validation
+    """
+    min_length: int = 8
+
+    """
+    Character sets used for checking minimum "complexity" in check_character_set validation
+    """
+    required_char_types: typing.List[PasswordCharset] = [
+        PasswordCharset(_('lowercase letter'), string.ascii_lowercase),
+        PasswordCharset(_('uppercase letter'), string.ascii_uppercase),
+        PasswordCharset(_('number'), string.digits),
+        PasswordCharset(_('symbol'), ''.join(sorted(
+            set(string.printable) -
+            set(string.whitespace + string.ascii_letters + string.digits))
+        )),
+    ]
+
+    """
+    Minimum character number of different character types required in check_character_set validation
+    """
+    required_min_char_types: int = 3
+
+    def check_length(self, pw: str, user):
+        """
+        Raises PasswordPolicyError if a password is not at least min_length characters long.
+        :param pw: password to check
+        :param user: user entity
+        """
+        if len(pw) < self.min_length:
+            raise PasswordPolicyError(_(
+                'Password must be at least {min_length} characters long',
+                min_length=self.min_length
+            ))
+
+    def check_character_set(self, pw: str, user):
+        """
+        Raises PasswordPolicyError if a password does not contain at least one character from at least
+        `required_at_least_char_types` of the alphabets in `required_char_sets`.
+        :param pw: password to check
+        :param user: user entity
+        """
+
+        def contains_char(pw, alphabet):
+            alphabet = set(alphabet)
+            return any(char in alphabet for char in pw)
+
+        missing = []
+        for name, alphabet in self.required_char_types:
+            if not contains_char(pw, alphabet):
+                missing.append(name)
+
+        if len(missing) > len(self.required_char_types) - self.required_min_char_types:
+            first_part = ', '.join(t.name for t in self.required_char_types[:-1])
+            raise PasswordPolicyError(_(
+                'Password must include at least {required} of {first} and/or {last}',
```

Comment:
It's not likely there would only be one character set required, but if there were this message might be confusing.